File: native-ios/MGLFleetSDK/Sources/MGLFleetSDK/FleetSdk.swift
import UIKit

public final class FleetSdk {
    public static let shared = FleetSdk()

    private var options: FleetSdkOptions?
    private let lock = NSLock()

    private init() {}

    /// Snapshot for internal callers (tests).
    func snapshotOptions() -> FleetSdkOptions? {
        lock.lock()
        defer { lock.unlock() }
        return options
    }

    public func initialize(options: FleetSdkOptions) {
        lock.lock()
        defer { lock.unlock() }
        self.options = options
    }

    /// Present fullscreen Fleet flow modally from `viewController`.
    public func presentFleetFlow(
        from viewController: UIViewController,
        session: FleetSessionOptions? = nil,
        completion: @escaping FleetSdkCompletionHandler,
    ) throws {
        lock.lock()
        let opts = options
        lock.unlock()

        guard opts != nil else {
            throw FleetSdkError(code: .notInitialized, message: "FleetSdk.initialize required.")
        }

        FleetPresentationBridge.pendingSession = session
        FleetPresentationBridge.pendingCompletion = completion

        let vc = FleetSdkViewController()
        vc.modalPresentationStyle = .fullScreen
        viewController.present(vc, animated: true)
    }

    public func addListener(tag: String, listener: FleetSdkEventListener) {
        FleetPresentationBridge.listeners.append((tag, listener))
    }

    public func removeListener(tag: String) {
        FleetPresentationBridge.listeners.removeAll { $0.0 == tag }
    }
}

File: native-android/fleet-sdk/src/main/java/com/mgl/fleet/sdk/FleetSdk.kt
package com.mgl.fleet.sdk

import android.content.Context
import android.content.Intent
import android.os.Looper
import androidx.fragment.app.FragmentActivity
import com.mgl.fleet.sdk.internal.FleetPresentationBridge
import com.mgl.fleet.sdk.internal.FleetSdkActivity

/**
 * Native Fleet SDK entry — initialize once, then present fullscreen flow.
 */
object FleetSdk {

    fun initialize(applicationContext: Context, options: FleetSdkOptions) {
        FleetSdkHolder.attach(applicationContext, options)
    }

    /** True once [initialize] has run successfully in this process. */
    fun isInitialized(): Boolean = FleetSdkHolder.isAttached()

    /**
     * Presents native Fleet UI over [activity]. Completion runs on main thread when flow finishes.
     */
    fun presentFleetFlow(
        activity: FragmentActivity,
        session: FleetSessionOptions?,
        callback: FleetSdkCompletionCallback,
    ) {
        FleetSdkHolder.applicationContext()
        FleetPresentationBridge.pendingSession = session
        FleetPresentationBridge.pendingCallback = callback
        val intent = Intent(activity, FleetSdkActivity::class.java)
        val launch = Runnable { activity.startActivity(intent) }
        if (Looper.myLooper() == Looper.getMainLooper()) {
            launch.run()
        } else {
            activity.runOnUiThread(launch)
        }
    }

    /** Optional analytics hooks — bridges may subscribe via JNI/reflection layers later. */
    fun addListener(tag: String, listener: FleetSdkEventListener) {
        FleetPresentationBridge.listeners.add(tag to listener)
    }

    fun removeListener(tag: String) {
        FleetPresentationBridge.listeners.removeAll { it.first == tag }
    }
}

fun interface FleetSdkEventListener {
    fun onEvent(name: String, payload: Map<String, Any?>?)
}
